File: src/examples/example_normal_channel/module/normal_publisher_module/normal_publisher_module.cc
#include "normal_publisher_module/normal_publisher_module.h"
#include "aimrt_module_cpp_interface/co/aimrt_context.h"
#include "aimrt_module_cpp_interface/co/inline_scheduler.h"
#include "aimrt_module_cpp_interface/co/on.h"
#include "aimrt_module_cpp_interface/co/schedule.h"
#include "aimrt_module_cpp_interface/co/sync_wait.h"
#include "aimrt_module_protobuf_interface/channel/protobuf_channel.h"
#include "aimrt_module_protobuf_interface/util/protobuf_tools.h"

#include "yaml-cpp/yaml.h"

namespace aimrt::examples::example_normal_channel::normal_publisher_module {

bool NormalPublisherModule::Initialize(aimrt::CoreRef core) noexcept {
  core_ = core;

  try {
    // Read cfg
    const auto configurator = core_.GetConfigurator();
    if (configurator) {
      YAML::Node cfg_node =
          YAML::LoadFile(std::string(configurator.GetConfigFilePath()));
      publish_topic_name_ = cfg_node["publish_topic_name"].as<std::string>();
      subscribe_topic_name_ = cfg_node["subscribe_topic_name"].as<std::string>();
      channel_frq_ = cfg_node["channel_frq"].as<double>();
    }

    // Get executor handle
    executor_ = core_.GetExecutorManager().GetExecutor("work_thread_pool");
    AIMRT_CHECK_ERROR_THROW(executor_ && executor_.SupportTimerSchedule(),
                            "Get executor 'work_thread_pool' failed.");

    // Register publish type
    publisher_ = core_.GetChannelHandle().GetPublisher(publish_topic_name_);
    AIMRT_CHECK_ERROR_THROW(publisher_, "Get publisher for topic '{}' failed.", publish_topic_name_);

    bool ret = aimrt::channel::RegisterPublishType<
        aimrt::protocols::example::ExampleEventMsg>(publisher_);
    AIMRT_CHECK_ERROR_THROW(ret, "Register publish type failed.");

    // 订阅事件
    subscriber_ = core_.GetChannelHandle().GetSubscriber(subscribe_topic_name_);
    AIMRT_CHECK_ERROR_THROW(
        subscriber_, "Get subscriber for topic '{}' failed.", subscribe_topic_name_);

    ret = aimrt::channel::SubscribeCo<aimrt::protocols::example::ExampleEventMsg>(
        subscriber_, std::bind(&NormalPublisherModule::EventHandle, this, std::placeholders::_1));
    AIMRT_CHECK_ERROR_THROW(ret, "Subscribe failed.");

  } catch (const std::exception& e) {
    AIMRT_ERROR("Init failed, {}", e.what());
    return false;
  }

  AIMRT_INFO("Init succeeded.");

  return true;
}

bool NormalPublisherModule::Start() noexcept {
  try {
    scope_.spawn(co::On(co::InlineScheduler(), MainLoop()));
  } catch (const std::exception& e) {
    AIMRT_ERROR("Start failed, {}", e.what());
    return false;
  }

  AIMRT_INFO("Start succeeded.");
  return true;
}

void NormalPublisherModule::Shutdown() noexcept {
  try {
    run_flag_ = false;
    co::SyncWait(scope_.complete());
  } catch (const std::exception& e) {
    AIMRT_ERROR("Shutdown failed, {}", e.what());
    return;
  }

  AIMRT_INFO("Shutdown succeeded.");
}

// Main loop
co::Task<void> NormalPublisherModule::MainLoop() {
  try {
    AIMRT_INFO("Start MainLoop.");

    co::AimRTScheduler work_thread_pool_scheduler(executor_);

    co_await co::Schedule(work_thread_pool_scheduler);

    uint32_t count = 0;
    while (run_flag_) {
      co_await co::ScheduleAfter(
          work_thread_pool_scheduler,
          std::chrono::microseconds(static_cast<uint32_t>(1000000 / channel_frq_)));
      count++;
      AIMRT_INFO("Loop count : {} -------------------------", count);

      // publish event
      aimrt::protocols::example::ExampleEventMsg msg;
      msg.set_msg("count: " + std::to_string(count));
      msg.set_num(count);
      AIMRT_INFO("Publish new pb event, data: {}",
                 aimrt::Pb2CompactJson(msg));
      aimrt::channel::Publish(publisher_, msg);
    }

    AIMRT_INFO("Exit MainLoop.");
  } catch (const std::exception& e) {
    AIMRT_ERROR("Exit MainLoop with exception, {}", e.what());
  }

  co_return;
}

co::Task<void> NormalPublisherModule::EventHandle(
    const aimrt::protocols::example::ExampleEventMsg& data) {
  AIMRT_INFO("Get new pb event, data: {}", aimrt::Pb2CompactJson(data));

  co_return;
}

}  // namespace aimrt::examples::example_normal_channel::normal_publisher_module

File: src/examples/py/protobuf_channel/py/normal_publisher_module.py
import aimrt_py
import aimrt_py_log
import aimrt_py_pb_chn
import yaml
import datetime
import time

from google.protobuf.json_format import MessageToJson
import event_pb2


class NormalPublisherModule(aimrt_py.ModuleBase):
    def __init__(self):
        super().__init__()
        self.core = aimrt_py.CoreRef()
        self.logger = aimrt_py.LoggerRef()
        self.work_executor = aimrt_py.ExecutorRef()

        self.topic_name = "test_topic"
        self.channel_frq = 0.5
        self.publisher = aimrt_py.PublisherRef()

        self.loop_count = 0
        self.stop_flag = False
        self.stoped_flag = False

    def Info(self):
        info = aimrt_py.ModuleInfo()
        info.name = "NormalPublisherModule"
        return info

    def Initialize(self, core):
        self.core = core
        self.logger = self.core.GetLogger()

        # log
        aimrt_py_log.info(self.logger, "Module initialize")

        try:
            # configure
            configurator = self.core.GetConfigurator()
            if (configurator):
                module_cfg_file_path = configurator.GetConfigFilePath()
                if (module_cfg_file_path):
                    with open(module_cfg_file_path, 'r') as file:
                        data = yaml.safe_load(file)
                        self.topic_name = str(data["topic_name"])
                        self.channel_frq = float(data["channel_frq"])

            # executor
            self.work_executor = self.core.GetExecutorManager().GetExecutor("work_thread_pool")
            if (not self.work_executor):
                aimrt_py_log.error(self.logger, "Get executor 'work_thread_pool' failed.")
                return False

            # channel-publisher
            self.publisher = self.core.GetChannelHandle().GetPublisher(self.topic_name)
            if (not self.publisher):
                aimrt_py_log.error(self.logger, "Get publisher for '{}' failed.".format(self.topic_name))
                return False

            aimrt_py_pb_chn.RegisterPublishType(self.publisher, event_pb2.ExampleEventMsg)

        except Exception as e:
            aimrt_py_log.error(self.logger, "Initialize failed. {}".format(e))
            return False

        return True

    def Start(self):
        aimrt_py_log.info(self.logger, "Module start")

        try:
            self.work_executor.Execute(self.PublishLoop)

        except Exception as e:
            aimrt_py_log.error(self.logger, "Initialize failed. {}".format(e))
            return False

        return True

    def Shutdown(self):
        self.stop_flag = True

        while (not self.stoped_flag):
            time.sleep(1)

        aimrt_py_log.info(self.logger, "Module shutdown")

    def PublishLoop(self):
        if (self.stop_flag):
            self.stoped_flag = True
            return

        try:
            self.loop_count = self.loop_count + 1
            aimrt_py_log.info(self.logger,
                              "Loop count : {} -------------------------".format(self.loop_count))

            # publish event
            event_msg = event_pb2.ExampleEventMsg()
            event_msg.msg = "count {}".format(self.loop_count)
            event_msg.num = self.loop_count
            aimrt_py_log.info(self.logger,
                              "Publish new pb event, data: {}".format(MessageToJson(event_msg)))

            aimrt_py_pb_chn.Publish(self.publisher, event_msg)

            # next loop
            self.work_executor.ExecuteAfter(
                datetime.timedelta(seconds=1 / self.channel_frq),
                self.PublishLoop)

        except Exception as e:
            aimrt_py_log.error(self.logger, "PublishLoop failed. {}".format(e))
            self.stoped_flag = True
